src/Grumplestiltskin/EllipticCurve.hs

```diff
@@ -0,0 +1,175 @@
+import Plutarch.Internal.PlutusType (pmatch)
+import Plutarch.Internal.Term (S, Term, punsafeCoerce)
+import Plutarch.Pair (PPair (PPair))
+import Plutarch.Prelude (
+    DeriveNewtypePlutusType (DeriveNewtypePlutusType),
+    PEq,
+    PInteger,
+    PPositive,
+    PlutusType,
+    pcon,
+    pif,
+    pupcast,
+    (#),
+    (#&&),
+    (#*),
+    (#+),
+    (#-),
+    (#==),
+ )
+
+{-
+Calculators:
+- "https://www.graui.de/code/elliptic2/"
+- "https://andrea.corbellini.name/ecc/interactive/modk-add.html"
+
+Resources:
+- "https://rareskills.io/post/elliptic-curves-finite-fields"
+- Handbook of Elliptic and Hyperelliptic Curve Cryptography -- Henri Cohen, ...
+- "http://koclab.cs.ucsb.edu/teaching/ecc/eccPapers/Washington-ch04.pdf" (some examples and proofs)
+- "https://andrea.corbellini.name/2015/05/17/elliptic-curve-cryptography-a-gentle-introduction/"
+
+Functionality of Elliptic curve over finite field. (cyclic groups)
+
+curve order = number of points on the curve (including point in infinity)
+field modulus = is the modulo we apply to the values
+
+## Functions
+
+-- TODO: representing the point at infinity?
+
+-- Doing P + Q
+-- This should be able to return point at infinity
+addPoints :: Point -> Point -> Point
+
+-- Doing 2P
+doublePoint :: Point -> Point
+
+-- P being the point
+-- k being the number of times we add `P` to itself. I.e. 3P = P + P + P
+-- this function calculates the kP
+mulPoint :: Integer -> Point -> Point
+
+-- Vertical symmetry
+-- x axis stays the same, and y is inversed using modulo, i.e. P.y + inv(P).y == field modulus
+invPoint :: Point -> Point
+
+## Tests
+Some tests may not be feasible since we don't know the curve's order.
+
+Associativity `(aP + bP) + cP == aP + (bP + cP)`
+Inverse `P + inv(P) = Identity`
+`doublePoint(P) == P + P`
+`xP == (x + k)P` when k is the order of the curve
+`(a + b % curve_order)P == aP + bP` (since we don't know curve order we should pick sufficiently small numbers)
+`invPoint P == P` when P is identity (in case we represent identity in some special way)
+
+## Notes
+
+From the resource:
+> What about optimized_bn128?
+
+Examining the library, you will see an implementation called optimized_bn128.
+If you benchmark the execution time, you will see this version runs much quicker,
+and it is the implementation used by pyEvm. For educational purposes however,
+it is preferable to use the non-optimized version as it structures the points in a more intuitive way (the usual x, y tuple).
+The optimized version structures EC points as 3-tuples, which are harder to interpret.
+
+13.2 section in Handbook of Elliptic and Hyperelliptic Curve Cryptography -- Henri Cohen
+-}
+
+newtype PECPoint (s :: S) = PECPoint (Term s (PPair PGFElement PGFElement))
+    deriving stock
+        (Generic)
+    deriving anyclass
+        (SOP.Generic)
+    deriving
+        (PlutusType)
+        via (DeriveNewtypePlutusType PECPoint)
```

Comment:
Why wrap `PPair` as a `newtype` here? If the goal is for this to be a `Data` encoded type, you probably want to make it a `data` with two fields, then use the struct encoding (which would serialize as a `BuiltinList`).